전남대 BE_나제법 4주차 과제(3단계)

nove1080 opened the pull request:
매번 코드리뷰를 남겨주셔서 감사합니다!
이번주 피드백 중 반영이 안된 부분은 확인 후에 현재 PR에 반영하도록 하겠습니다

## 구현 목록
- [x] 상품 도메인 테스트 작성
- [x] 상품 옵션 재고 차감기능 구현
- [x] 상품 옵션 도메인 테스트 작성

nove1080 committed:
init: 3주차 코드 업로드

nove1080 committed:
docs: README.md 4주차 1단계 요구사항 반영

nove1080 committed:
remove: ErrorRespondable.java

nove1080 committed:
remove: schema.sql

nove1080 committed:
refactor: 예외 코드 롤백

- ErrorRespondable -> CustomException 사용하는 코드로 변경

nove1080 committed:
feat: data.sql 추가

nove1080 committed:
style: SpecialCharacter.java 불필요한 줄바꿈 제거

nove1080 committed:
feat: HexColor.java 추가

nove1080 committed:
feat: HexColorValidator.java 추가

nove1080 committed:
feat: Category.java 추가

nove1080 committed:
feat: CategoryRepository.java 추가

nove1080 committed:
feat: Color.java 추가

nove1080 committed:
feat: CategoryService.java 추가

nove1080 committed:
feat: CreateCategoryRequest.java 추가

nove1080 committed:
feat: CreateCategoryResponse.java 추가

nove1080 committed:
feat: ReadCategoryResponse.java 추가

nove1080 committed:
feat: ReadAllCategoriesResponse.java 추가

nove1080 committed:
feat: Product.java Category 필드 추가

nove1080 committed:
style: CreateMemberRequest.java 줄바꿈 추가

nove1080 committed:
feat: CreateProductRequest.java categoryId 필드 추가

nove1080 committed:
feat: ProductApiController.java 단일 상품 조회 API 추가

nove1080 committed:
feat: CategoryApiController.java 추가

nove1080 committed:
feat: ProductService.java createProduct() 카테고리 정보 추가

nove1080 committed:
feat: ReadProductResponse.java 카테고리 정보 추가

nove1080 committed:
feat: UpdateProductResponse.java 카테고리 정보 추가

nove1080 committed:
feat: CategoryService.java readCategory() 추가

nove1080 committed:
feat: CategoryApiController.java 단일 카테고리 조회 API 추가

nove1080 committed:
feat: data.sql 카테고리 더미 데이터 추가

nove1080 committed:
feat: CategoryService.java readAllCategories 페이징 처리

nove1080 committed:
feat: CategoryApiController.java 전체 카테고리 조회 API 페이징 처리

nove1080 committed:
test: CategoryDummyDataProvider.java 추가

nove1080 committed:
test: ProductDummyDataProvider.java Category 컬럼 추가

nove1080 committed:
test: MemberApiControllerTest.java CategoryDummyDataProvider 의존성 추가

nove1080 committed:
feat: ProductRepository.java findByCategoryId 추가

nove1080 committed:
feat: ProductService.java readProductsByCategoryId 추가

nove1080 committed:
refactor: ProductViewController.java

nove1080 committed:
feat: ProductApiController.java 카테고리 아이디로 상품 조회 API 추가

nove1080 committed:
style: ProductApiController.java 메서드 선언 순서 변경

nove1080 committed:
feat: ErrorCode.java HttpStatus 필드 추가

nove1080 committed:
rename: ErrorCode.java -> ErrorStatus.java

nove1080 committed:
refactor: ApiExceptionHandler.java Http 상태 코드 동적으로 반환

nove1080 committed:
refactor: BaseEntity, BaseTimeEntity 패키지 위치 변경

nove1080 committed:
feat: UpdateCategoryRequest.java 추가

nove1080 committed:
feat: UpdateCategoryResponse.java 추가

nove1080 committed:
feat: Category.java update() 추가

nove1080 committed:
feat: CategoryService.java 수정, 삭제 기능 구현

nove1080 committed:
fix: CategoryService.java 트랜잭션 적용

nove1080 committed:
feat: CategoryApiController.java 카테고리 수정, 삭제 API 구현

nove1080 committed:
refactor: ErrorCode -> ErrorStatus 이름 변경

nove1080 committed:
test: CategoryServiceTest.java 추가

nove1080 committed:
feat: Member.java WishProduct 와의 연관관계 제거

nove1080 committed:
feat: Product.java WishProduct 와의 연관관계 제거

nove1080 committed:
feat: WishProductRepository.java deleteAllByProductId, MemberId 추가

nove1080 committed:
feat: ProductService.java deleteProduct() 위시 상품 제거 로직 추가

nove1080 committed:
feat: MemberService.java deleteMember() 위시 상품 제거 로직 추가

nove1080 committed:
refactor: MemberDetailsService.java 트랜잭션 어노테이션 적용

nove1080 committed:
remove: ClientException.java, ServerException.java

nove1080 committed:
refactor: 커스텀 예외 부모 클래스 변경

nove1080 committed:
refactor: 커스텀 예외 핸들러 통합

nove1080 committed:
Merge branch 'nove1080' into step1

nove1080 committed:
docs: README.md 4주차 2단계 요구사항 반영

nove1080 committed:
remove: 중복 코드 제거

nove1080 committed:
feat: ProductOption.java 추가

nove1080 committed:
feat: ReadProductOptionResponse.java 추가

nove1080 committed:
feat: ReadAllProductOptionsResponse.java 추가

nove1080 committed:
feat: ProductOptionRepository.java 추가

nove1080 committed:
feat: ProductOptionService.java 추가

nove1080 committed:
feat: ProductApiController.java 상품 옵션 조회 API 구현

nove1080 committed:
feat: data.sql 상품 옵션 더미데이터 추가

nove1080 committed:
feat: CreateProductOptionRequest.java 추가

nove1080 committed:
feat: CreateProductOptionResponse.java 추가

nove1080 committed:
remove: 사용하지 않는 클래스 삭제

- ClientException
- ErrorCode
- ErrorRespondable
- ServerException

nove1080 committed:
feat: BadRequestException.java 기본 생성자 제거

nove1080 committed:
feat: ErrorStatus.java ALREADY_EXISTS 추가

nove1080 committed:
feat: CreateProductOptionRequest.java toEntity() 추가

nove1080 committed:
feat: CreateProductOptionResponse.java fromEntity() 추가

nove1080 committed:
feat: Product.java ProductOption에 대한 참조 추가

nove1080 committed:
feat: ProductOption.java Product에 대한 참조 제거

- Product의 Id만 참조하도록 수정

nove1080 committed:
feat: ProductOptionRepository.java existsByNameAndProductId() 추가

nove1080 committed:
feat: ProductOptionService.java createOption() 추가

nove1080 committed:
feat: ProductApiController.java 상품 옵션 등록 API 구현

nove1080 committed:
feat: AlreadyExistsException.java 추가

nove1080 committed:
fix: ProductOptionService.java @Transactional 적용

nove1080 committed:
feat: application-local.yml 토큰 유지 기간 1시간 -> 24시간 으로 변경

nove1080 committed:
feat: ResourceNotFoundException.java 추가

nove1080 committed:
feat: ProductOptionService.java deleteOption() 추가

nove1080 committed:
feat: ProductApiController.java 상품 옵션 삭제 API 구현

nove1080 committed:
feat: ProductOption.java update() 추가

nove1080 committed:
feat: ProductOptionRepository.java findDuplicatedProductOption 추가

nove1080 committed:
feat: UpdateProductOptionRequest.java 추가

nove1080 committed:
feat: UpdateProductOptionResponse.java 추가

nove1080 committed:
feat: ProductOptionService 옵션 전체 조회, 옵션 수정 기능 구현

nove1080 committed:
feat: ProductApiController.java 상품 수정 API 구현

nove1080 committed:
refactor: ProductApiController.java id 대신 productId 으로 변수명 변경

nove1080 committed:
refactor: ProductOptionService.java 로그 삭제

nove1080 committed:
refactor: ProductOptionService.java validateOptionNameExists 오버로딩

nove1080 committed:
style: ProductOptionService.java validateOptionNameExists 선언 위치 변경

nove1080 committed:
feat: Product.Builder productOptions 추가

nove1080 committed:
feat: CreateProductRequest.java 상품 옵션 필드 추가

nove1080 committed:
feat: ProductOptionRepository.java deleteAllByProductId 추가

nove1080 committed:
feat: ProductService.java 상품 추가 시, option도 같이 저장하도록 로직 수정

nove1080 committed:
feat: ProductService.java 상품 삭제 시, 옵션도 같이 삭제 되도록 로직 수정

nove1080 committed:
feat: ProductOptionService.java deleteAllOptionsByProductId 추가

nove1080 committed:
feat: CreateProductOptionRequest.java toEntity() 추가

nove1080 committed:
test: ProductServiceTest.java 추가

nove1080 committed:
refactor: CreateProductRequest.java URL 필드 타입 변경

- URL -> String

nove1080 committed:
feat: CreateProductResponse.java 필드 추가

nove1080 committed:
test: ProductOptionServiceTest.java 추가

nove1080 committed:
feat: CreateMemberResponse.java email, name 필드 추가

nove1080 committed:
test: MemberServiceTest.java 추가

nove1080 committed:
feat: CreateWishProductResponse.java quantity 필드 추가

nove1080 committed:
test: WishProductServiceTest.java 추가

nove1080 committed:
feat: Product.java 생성자 productOptions 초기화 추가

nove1080 committed:
refactor: UpdateProductRequest.java imageUrl 타입 URL -> String 변경

nove1080 committed:
refactor: CreateProductRequest.java getProductOptions() 수정

nove1080 committed:
test: ProductServiceTest.java 추가

nove1080 committed:
test: ProductServiceTest.java wishProductRepository 목 객체 추가

nove1080 committed:
feat: Product update 메서드 파라미터 수정

- 파라미터가 객체인 경우 Builder.build()의 검증 로직으로 인해 업데이트에 필요없는 데이터를 채워야하는 문제를 개선하기 위함

nove1080 committed:
feat: ProductService.java 상품 업데이트 로직 수정

- Product update 메서드 파라미터 수정으로 인한 변경

nove1080 committed:
refactor: UpdateProductResponse.java ImageUrl 타입 변경

- URL -> String

nove1080 committed:
test: ProductServiceTest.java updateProduct 테스트 추가

nove1080 committed:
docs: README.md 4주차 3단계 요구사항 반영

nove1080 committed:
feat: ProductOption.java subtractQuantity() 추가

nove1080 committed:
test: ProductTest.java 추가

nove1080 committed:
refactor: Product.java 검증 로직 분리

nove1080 committed:
feat: ProductOption.java 재고 차감 요청에 대한 검증 추가

nove1080 committed:
feat: ProductOptionService.java subtractOptionStock() 추가

nove1080 committed:
feat: SubtractProductOptionQuantityRequest.java 추가

nove1080 committed:
feat: SubtractProductOptionQuantityResponse.java 추가

nove1080 committed:
test: ProductOptionServiceTest.java subtractOptionStock 테스트 추가

nove1080 committed:
Merge branch 'nove1080' into step3

nove1080 committed:
Merge branch 'nove1080' into step3

Hongdonggeon commented on src/main/java/gift/domain/validator/ProductValidator.java:
```diff
@@ -0,0 +1,14 @@
+
+public abstract class ProductValidator {
+
+    public static void validateProductOptionsPresence(List<ProductOption> productOptions) {
```
사용되고 있지 않은 것 같네요

nove1080 replied:
`Product`의 `Builder`에서 사용하고 있었는데 이 부분은 동건님의 말씀대로 별도의 Validator 클래스를 만들지 않고 도메인에 검증코드를 작성하는 방법으로 수정하겠습니다

Hongdonggeon commented on src/main/java/gift/domain/validator/ProductOptionValidator.java:
```diff
@@ -0,0 +1,11 @@
+
+public abstract class ProductOptionValidator {
+
+    public static void validatedRequestQuantity(int stock, int quantity) {
```
validator 클래스를 따로 구분할 필요 없이, ProductOption 엔티티에서 이 메서드를 만들어도 될 것 같아요

nove1080 replied:
이후에 검증할 부분이 많아지면 코드가 길어지는게 아닐까 라는 생각에서 validator를 따로 생성하였는데 
어짜피 도메인 검증 코드는 해당 도메인 안에서만 사용되는데 따로 분리하는 바람에 이것이 도메인 클래스의 응집도를 떨어뜨리게 되어서 이러한 이유에서 엔티티 안에 검증 로직을 넣어두려는데 다른 이유가 더 있을까요? 

혹시 추가로 궁금한 건데 검증 코드가 많아지더라도 도메인 안에 위치하는게 더 나은가요? 혹은 다른 방법이 있을까요?

Hongdonggeon replied:
검증 로직이 많아지면 많아질수록 각자 도메인이 나눠가지는게 응집도가 더좋지 않을까여?
validator에서 모이게 된다면 오히려 validator가 가지는 역할이 너무 많아질 것 같아요

Hongdonggeon reviewed (changes requested):
안녕하세요 제법님
요구사항이 간단하네요 코멘트 남겨 놓았습니다! 확인 한번 해주세요

nove1080 committed:
remove: 도메인 검증기 삭제

- ProductValidator.java
- ProductOptionValidator.java
- 응집도를 높이기 위해 도메인에 검증 코드를 작성

nove1080 committed:
feat: ProductOption.java 수량 차감에 대한 검증 로직 추가

nove1080 committed:
refactor: Product.java 상품 옵션에 대한 검증 로직 수행 위치 변경 및 검증 로직 직접 선언

nove1080 committed:
feat: ProductOptionService.java createInitialOptions() 추가

nove1080 committed:
refactor: ProductService.java 상품 등록 시 호출하는 상품 옵션 저장 메서드 변경

nove1080 committed:
test: ProductOptionServiceTest.java createInitialOptions 테스트 추가

nove1080 committed:
test: ProductServiceTest.java 상품 생성 실패 테스트 추가

nove1080 committed:
Merge remote-tracking branch 'origin/step3' into step3

nove1080 commented:
안녕하세요 동건님 
1. 도메인 검증 로직을 해당 도메인 안에서 구현
2. 2단계에서 남기신 https://github.com/kakao-tech-campus-2nd-step2/spring-gift-enhancement/pull/329#pullrequestreview-2190689209 옵션 생성 시 불필요한 조회 쿼리 발생 없애기

위 두 가지 사항을 반영하였습니다!

옵션 생성에 관한 구현은 좀 더 나은 구현 방법이 있을 것 같아 조금 더 고민해보겠습니다

Hongdonggeon reviewed (approved):
안녕하세요 제법님
3단계 미션 머지할게요
질문에 대한 답변도 확인해주세요
